style_guides/angular_style_guide.md

```diff
@@ -63,3 +63,179 @@ app.controller('myController', function($scope, otherDeps, Private) {
 
 **Note:** Use this sparingly. Whenever possible, your modules should not be
 coupled to the angular application itself.
+
+## Directives
+
+### Use callbacks to communicate side effects up the hierarchy
+
+  - Treat attribute bindings as read-only and communicate with with the parent
+    directive via callbacks. Essentially treat directive attribute bindings
+    like function arguments, which should also not be mutated.
+
+    *Why?*: Spreading mutable state throughout the application makes it very
+    difficult to understand and predict. By receiving data from the parent
+    through the bindings and communicating desired changes back via separate
+    callback attributes the data flow becomes explicit. The attribute bindings
+    then become the public interface of the component, describing all the
+    inputs dictating the rendering and all possible effects on the application
+    state.
+
+  ```javascript
+  /* avoid */
+  module.directive('myBadTableHeader', function MyBadTableHeader() {
+    return {
+      bindToController: true,
+      controllerAs: 'tableHeader',
+      controller: TableHeaderController,
+      replace: true,
+      scope: {
+        caption: '=',
+        sortColumn: '=',
+      },
+      template: `<th
+        ng-class="{ 'tableHeader--sortColumn': tableHeader.isSortColumn() }"
+        ng-bind="tableHeader.caption"
+        ng-click="tableHeader.sort()"
+      ></th>`,
+    };
+  });
+
+  function TableHeaderController() {
+    this.isSortColumn = () => (
+      this.sortColumn === this.caption
+    );
+
+    this.sort = () => (
+      this.sortColumn = this.caption  // AVOID: mutating input attributes
+    );
+
+    /* OTHER STUFF */
+  }
+  ```
+
+  ```javascript
+  /* recommended */
+  module.directive('myGoodTableHeader', function MyGoodTableHeader() {
+    return {
+      bindToController: true,
+      controllerAs: 'tableHeader',
+      controller: TableHeaderController,
+      replace: true,
+      scope: {
+        caption: '=',          // INPUT
+        isSortColumn: '=',     // INPUT
+        onChangeSorting: '=',  // EFFECT
```

Comment:
After reading the next section, I have another bullet point:

- using expressions rather than binding function allows intermediate directives (between the top level directive that contains the majority of the business logic and the leaf directive that calls the handler/callback/etc) to define the API it exposes rather than relying on the different child elements it consumes to expose the proper API. Continuing with the header example:
    - If the table directive exposes a `onFilterClick: '='` effect which is then passed to the header and again passed to each cell, the header doesn't necessarily know how the cell is going to call the `onFilterClick` effect, or even if there is an `onFilterClick` effect to be called. It is instead the responsibility of each cell to consume this effect properly, meaning the cells ultimately define the API for the table and table header directives.
    - Alternatively, the table, header and cell could all implement the `onFilterClick: '&'` effect and at each level we would define how we consume the behavior exposed by the direct child... the table would define
       ```html
       <th my-good-table-header on-filter-click="table.onFilterClick($locals)">
       ```
       and the table header would define
       ```html
       <td my-good-table-cell on-filter-click="tableHeader.onFilterClick($locals)">
       ```
       effectively proxying the `onFilterClick()` calls all the way from the cell to the table. If needed though, the header could implement different logic for different `*-table-cell` directives, ensuring that the `my-good-table-header` is in control of the API that it exposes.